### DISCUSSION TITLE
NumberField + useController React Hook Form

### DISCUSSION THREAD

**Discussion by cojoclaudiu** [adobe/react-spectrum#4484] [General]
Hi, I'm this example for my component and I'm trying to make a controlled component using React Hook form but when the input is empty and I click outside I get a 0 value inside the input instead of empty input.

Is there any way to fix this?
Demo: https://codesandbox.io/s/usecontroller-forked-viw53w?file=/src/App.js

```ts
function ControlledNumberField<TFieldValues extends FieldValues = FieldValues>({
	name,
	control,
	defaultValue,
	...props
}: ControlledNumberFieldProp<TFieldValues>) {
	const {
		field: { value, ...field },
		fieldState: { error },
	} = useController({ name, defaultValue, control });

	return (
		<NumberField
			aria-describedby={props.label}
			aria-invalid={Boolean(error)}
			errorMessage={error?.message}
			value={value}
			{...field}
			{...props}
		/>
	);
}```

**Comment by snowystinger** [comment]
We don't fire an event if it's empty and you click in and back outside https://codesandbox.io/s/magical-lake-zrslyr?file=/src/App.js so I would look and see if React Hook Form is doing something. Maybe related to https://github.com/react-hook-form/react-hook-form/issues/6287 ?

Otherwise, you could make a codesandbox here reproducing the issue with explicit steps on how to interact with the UI to cause the bug. Maybe I missed a step just now.

**Reply by cojoclaudiu** [reply]
Hi, thank you for replay. I will be back with my codesandbox soon

**Reply by cojoclaudiu** [reply]
@snowystinger here is my actual input setup (when you add the numbers inside the input -> delete all -> click outside the input = 0):
[codesandbox updated](https://codesandbox.io/s/usecontroller-forked-viw53w?file=/src/App.js)

**Reply by snowystinger** [reply]
So that link isn't loading for me.
I tried those steps on our own NumberField in the codesandbox I shared, I'm still not getting an onChange for
1. click on NumberField
2. Enter numbers
3. Delete numbers
4. click outside NumberField

**Reply by cojoclaudiu** [reply]
> So that link isn't loading for me. I tried those steps on our own NumberField in the codesandbox I shared, I'm still not getting an onChange for
> 
>     1. click on NumberField
> 
>     2. Enter numbers
> 
>     3. Delete numbers
> 
>     4. click outside NumberField
@snowystinger thank you for the replay. I've updated the codesandbox link, hope now it's working. The input doesn't work when it's used with React Hook Form, useControlled Hook 

https://noc686-5173.csb.app/
https://codesandbox.io/p/github/cojoclaudiu/vite-ts-starter/Input-Number-React-Aria-%252B-RHF

**Reply by snowystinger** [reply]
I apologize, I'm thoroughly confused by the codesandboxes you're linking. For instance, I don't even see useController in any of the source files. Can you make a simpler one? Here's one I copied from React Hook Forms and added our NumberField to. I probably haven't hooked everything up correctly since I've not used RHF before. But it should be a starting place to make a minimal reproduction.
https://react-hook-form.com/api/usecontroller/#example
https://codesandbox.io/s/usecontroller-forked-12qs6n?file=/src/App.js

**Reply by cojoclaudiu** [reply]
@snowystinger hi, sorry for the confusion (I don't know what's happen with the new codesandbox). I've updated your codesandox with my setup and I get the same behaviour https://codesandbox.io/s/usecontroller-forked-viw53w?file=/src/App.js

Thank you for help.

**Reply by snowystinger** [reply]
Ah, right, so one issue is that value can only be of type number, so to pass no number, you can use `NaN`

I've addressed it in your codesandbox using a combination of these pages of documentation and discussion and here's the working sandbox
https://codesandbox.io/s/usecontroller-forked-qbginj?file=/src/App.js

https://github.com/react-hook-form/react-hook-form/issues/6287
https://github.com/orgs/react-hook-form/discussions/8068
https://react-spectrum.adobe.com/react-spectrum/NumberField.html#props

**Reply by cojoclaudiu** [reply]
@snowystinger thank you for help, I've fixed the issue based on your reply

Later edit: actually not working
```ts
function ControlledNumberField<TFieldValues extends FieldValues = FieldValues>({
	name,
	control,
	defaultValue,
	...props
}: ControlledNumberFieldProp<TFieldValues>) {
	const {
		field: { value, ...field },
		fieldState: { error },
	} = useController({ name, defaultValue, control });

	return (
		<NumberField
			aria-describedby={props.label}
			aria-invalid={Boolean(error)}
			value={+(value || NaN)}
			errorMessage={error?.message}
			defaultValue={+(defaultValue || NaN)}
			{...field}
			{...props}
		/>
	);
}
```

**Reply by snowystinger** [reply]
I think your `...field` is overwriting the `value` and `defaultValue` you've specified, you'd need to do it in this order instead
```
			errorMessage={error?.message}
			{...field}
			{...props}
			value={+(value || NaN)}
			defaultValue={+(defaultValue || NaN)}
```

**Reply by cojoclaudiu** [reply]
@snowystinger I think the values is not passed to the `inputValue` from `useNumberFieldState`. If I pass the value from the controlled to the inputValue than I got the correct input, but this will brake the zod validation

**Reply by snowystinger** [reply]
I've stripped out most everything in my sandbox copy https://codesandbox.io/s/usecontroller-forked-qbginj?file=/src/App.js
I'm not sure you need the lines `value={+(value || NaN)}` or `defaultValue={+(defaultValue || NaN)}` or even the override to onChange I provided before.

Please let me know if this version has your issue. Otherwise I'll need a sandbox with the minimum setup and clearly defined steps on how to reproduce with expectations at each step vs reality. I apologize if I'm misunderstanding you. Thank you for your understanding.
